### PR TITLE
Refactoring of x_columns and aml_features

### PR DESCRIPTION
**Pull request by nahuel-coliva**

### CODE REVIEW

File: check.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import os
+import sys
+import argparse
+
+parser = argparse.ArgumentParser(description="Performs fault tolerance tests")
+parser.add_argument('-t', "--timeout", help="time elapsed between interruptions (in seconds)", type=float, default=10)
+parser.add_argument('-o', "--output", help="output folder where all the models will be stored", default="output_fault_tolerance")
+args = parser.parse_args()
+
+print(args)
```
Comment (brunoguindani) [inline check.py]: Missing newline at the end of the file

File: model_building/experiment_configuration.py [hunk 2]
```diff
@@ -73,7 +76,7 @@ class ExperimentConfiguration(abc.ABC):
     ----------
     _campaign_configuration: dict of str: dict of str: str
         The set of options specified by the user though command line and campaign configuration files
-
+
```
Comment (brunoguindani) [inline model_building/experiment_configuration.py]: Spaces to be removed

File: model_building/experiment_configuration.py [hunk 3]
```diff
@@ -213,7 +216,7 @@ def __init__(self, campaign_configuration, hyperparameters, regression_inputs, p
             if not os.path.exists(self._experiment_directory):
                 os.makedirs(self._experiment_directory)
 
-    def train(self, force=False):
+    def train(self, force=False): #(self, model_ID, force=False)
```
Comment (brunoguindani) [inline model_building/experiment_configuration.py]: What is this comment? Please remove this if not needed

File: model_building/experiment_configuration.py [hunk 4]
```diff
@@ -225,12 +228,19 @@ def train(self, force=False):
...
-                    self.set_regressor(pickle.load(f))
-                self.trained = True
+                    regressor_obj = pickle.load(f) #keep
+                    if force: #re-training the model requires keeping the same hyperparameters previously found
+                        self._hyperparameters = regressor_obj.get_hypers()
+                    else:
+                        self.set_regressor(regressor_obj.get_regressor())
+                        self.set_x_columns(regressor_obj.get_x_columns())
+                        self._hyperparameters = regressor_obj.get_hypers() #possibly not needed
+                        self.trained = True
```
Comment (brunoguindani) [inline model_building/experiment_configuration.py]: This block should be outside the scope of `with()`, i.e. indented one less time

File: model_building/experiment_configuration.py [hunk 5]
```diff
@@ -241,11 +251,15 @@ def train(self, force=False):
             warnings.simplefilter("ignore")
             self._train()
         self.trained = True
+        """
         if self._regressor and not hasattr(self._regressor, 'aml_features'):
             self._regressor.aml_features = self.get_x_columns()
+        """
```
Comment (brunoguindani) [inline model_building/experiment_configuration.py]: Please remove these lines before merging

File: model_building/lr_ridge_experiment_configuration.py [hunk 6]
```diff
@@ -97,11 +97,25 @@ def print_model(self):
...
         ret_string = initial_string
         coefficients = self._regressor.coef_
-        assert len(self._regressor.aml_features) == len(coefficients)
+
+        #TODO: togli il try-except, lascia solo l'assert
+        try:
+            assert len(self.get_x_columns()) == len(coefficients) #len(self._regressor.aml_features) == len(coefficients)
+        except:
+            print("Assertion error:",str(self.get_x_columns()),str(coefficients),sep=" ")
+            #sys.exit(1)
```
Comment (brunoguindani) [inline model_building/lr_ridge_experiment_configuration.py]: :)

File: model_building/lr_ridge_experiment_configuration.py [hunk 7]
```diff
@@ -97,11 +97,25 @@ def print_model(self):
...
         # Show coefficients in order of decresing absolute value
         idxs = np.argsort(np.abs(coefficients))[::-1]
+
         for i in idxs:
-            column = self._regressor.aml_features[i]
+            try:
+                column = self.get_x_columns()[i] #self._regressor.aml_features[i]
+            except Exception as e:
+                print(e)
+                continue
```
Comment (brunoguindani) [inline model_building/lr_ridge_experiment_configuration.py]: What's the purpose of this?

File: model_building/model_building.py [hunk 8]
```diff
@@ -125,7 +126,7 @@ def process(self, campaign_configuration, regression_inputs, processes_number):
                     exp.train()
                 else:
                     try:
-                        exp.train()
+                        exp.train()
```
Comment (brunoguindani) [inline model_building/model_building.py]: Space to be removed

File: model_building/model_building.py [hunk 9]
```diff
@@ -167,7 +175,7 @@ def process(self, campaign_configuration, regression_inputs, processes_number):
         for technique in best_confs:
             best_conf = best_confs[technique]
             # Get information about the used x_columns
-            all_data.x_columns = best_conf.get_regressor().aml_features
+            all_data.x_columns = best_conf.get_x_columns() #get_regressor().aml_features
```
Comment (brunoguindani) [inline model_building/model_building.py]: Comment to be removed

File: model_building/model_building.py [hunk 10]
```diff
@@ -206,7 +221,11 @@ def process(self, campaign_configuration, regression_inputs, processes_number):
             self._logger.removeHandler(file_handler)
 
             # Build the regressor with all data
-            best_regressors[technique] = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), all_data.scalers)
+            if hasattr(best_conf, '_wrapped_experiment_configuration'):
+                    best_regressors[technique] = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), all_data.scalers, best_conf._wrapped_experiment_configuration._hyperparameters)
+            else:
+                best_regressors[technique] = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), all_data.scalers, best_conf._hyperparameters)
```
Comment (brunoguindani) [inline model_building/model_building.py]: See above

File: model_building/nnls_experiment_configuration.py [hunk 11]
```diff
@@ -97,11 +97,11 @@ def print_model(self):
         initial_string = "NNLS coefficients:\n"
         ret_string = initial_string
         coefficients = self._regressor.coef_
-        assert len(self._regressor.aml_features) == len(coefficients)
+        assert len(self.get_x_columns()) == len(coefficients) #len(self._regressor.aml_features) == len(coefficients)
         # Show coefficients in order of decresing absolute value
         idxs = np.argsort(np.abs(coefficients))[::-1]
         for i in idxs:
-            column = self._regressor.aml_features[i]
+            column = self.get_x_columns()[i] #self._regressor.aml_features[i]
```
Comment (brunoguindani) [inline model_building/nnls_experiment_configuration.py]: Comments to be removed. Also, doing `columns = self.get_x_columns()` and then calling that variable is a little bit more elegant

File: model_building/wrapper_experiment_configuration.py [hunk 12]
```diff
@@ -148,7 +155,15 @@ def get_x_columns(self):
...
-        return self._wrapped_experiment_configuration.get_x_columns()
+        x_cols = super().get_x_columns() #self._wrapped_experiment_configuration.get_x_columns()
+        """
+        try:
+            assert self._wrapped_experiment_configuration.get_x_columns() == super().get_x_columns()
+        except:
+            super().set_x_columns(x_cols)
+        """
+        return x_cols
+        #return super().get_x_columns()
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: Comments to be removed

File: model_building/wrapper_experiment_configuration.py [hunk 13]
```diff
@@ -159,8 +174,27 @@ def set_x_columns(self, x_cols):
...
         super().set_x_columns(x_cols)
+        #self._logger.info("Wrapper:")
+        #self._logger.info(self._regression_inputs.__str__())
+
+        #self._logger.info("Wrapped:")
         self._wrapped_experiment_configuration.set_x_columns(x_cols)
+        #self._logger.info(self._wrapped_experiment_configuration._regression_inputs.__str__())
+
+        #self._logger.info("Get x_columns:")
+        #self._logger.info(str(self.get_x_columns()))
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: Comments to be removed

File: model_building/wrapper_experiment_configuration.py [hunk 14]
```diff
@@ -234,7 +268,7 @@ def _train(self):
         filtered_xdata = self._sfs.transform(xdata)  # is an np.array
         filtered_xdata = pd.DataFrame(filtered_xdata, columns=x_cols)
         self.set_x_columns(x_cols)
-        self.get_regressor().aml_features = x_cols
+        #self.get_regressor().aml_features = x_cols
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: :^)

File: model_building/wrapper_experiment_configuration.py [hunk 15]
```diff
@@ -262,7 +297,8 @@ def print_model(self):
         """
         Print the representation of the generated model
         """
-        return "".join(("Features selected with SFS: ", str(self.get_regressor().aml_features), "\n",
+        #return "".join(("Features selected with SFS: ", str(self.get_regressor().aml_features), "\n",
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: !

File: model_building/wrapper_experiment_configuration.py [hunk 16]
```diff
@@ -420,7 +456,7 @@ def _train(self):
         self._wrapped_experiment_configuration._regressor = self._wrapped_regressor
         self._wrapped_experiment_configuration._hyperparameters = best_param
         self._wrapped_experiment_configuration._train()
-        self.get_regressor().aml_features = self.get_x_columns()
+        #self.get_regressor().aml_features = self.get_x_columns()
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: 🅱️

File: results.py [hunk 17]
```diff
@@ -98,6 +98,24 @@ def collect_data(self):
         Collect the data of all the performed experiments
         """
         exp_conf: ec.ExperimentConfiguration
+
+        #Remove
```
Comment (brunoguindani) [inline results.py]: \+

File: model_building/model_building.py [hunk 18]
```diff
@@ -184,7 +192,11 @@ def process(self, campaign_configuration, regression_inputs, processes_number):
 
             if 'save_training_regressors' in campaign_configuration['General'] and campaign_configuration['General']['save_training_regressors']:
                 # Build the regressor trained on the train set only
-                training_regressor = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), regression_inputs.scalers)
+                if hasattr(best_conf, '_wrapped_experiment_configuration'):
+                    training_regressor = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), regression_inputs.scalers, best_conf._wrapped_experiment_configuration._hyperparameters)
+                else:
+                    training_regressor = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), regression_inputs.scalers, best_conf._hyperparameters)
+
```
Comment (brunoguindani) [inline model_building/model_building.py]: For the sake of clarity, I'd rather do something like:
```py
hypers = best_conf._wrapped_experiment_configuration._hyperparameters if hasattr(best_conf, '_wrapped_experiment_configuration') else best_conf._hyperparameters
training_regressor = regressor.Regressor(..., hypers)
```

File: model_building/model_building.py [hunk 19]
```diff
@@ -184,7 +192,11 @@ def process(self, campaign_configuration, regression_inputs, processes_number):
 
             if 'save_training_regressors' in campaign_configuration['General'] and campaign_configuration['General']['save_training_regressors']:
                 # Build the regressor trained on the train set only
-                training_regressor = regressor.Regressor(campaign_configuration, best_conf.get_regressor(), best_conf.get_x_columns(), regression_inputs.scalers)
+                if hasattr(best_conf, '_wrapped_experiment_configuration'):
```
Comment (brunoguindani) [inline model_building/model_building.py]: Given how often it appears, I'd create a `is_wrapper()` class method (or other names you like better) that performs this `hasattr()` call

**Review by brunoguindani — COMMENTED** [review]
There appears to be a conflict in the LRRidge file, please fix it

File: check.py (hunk 1 above)
Comment (nahuel-coliva) [reply check.py]: Removed file entirely

File: model_building/experiment_configuration.py (hunk 2 above)
Comment (nahuel-coliva) [reply model_building/experiment_configuration.py]: Removed

File: model_building/experiment_configuration.py (hunk 3 above)
Comment (nahuel-coliva) [reply model_building/experiment_configuration.py]: model_ID was a variable used during the testing, forgot to remove this comment as well as the ones from the other issues, now fixed

File: model_building/experiment_configuration.py (hunk 4 above)
Comment (nahuel-coliva) [reply model_building/experiment_configuration.py]: Unindented

File: model_building/experiment_configuration.py (hunk 5 above)
Comment (nahuel-coliva) [reply model_building/experiment_configuration.py]: Removed

File: model_building/lr_ridge_experiment_configuration.py (hunk 6 above)
Comment (nahuel-coliva) [reply model_building/lr_ridge_experiment_configuration.py]: (:

File: model_building/lr_ridge_experiment_configuration.py (hunk 7 above)
Comment (nahuel-coliva) [reply model_building/lr_ridge_experiment_configuration.py]: Receiving the exception that was causing the program to break, now removed

File: model_building/model_building.py (hunk 8 above)
Comment (nahuel-coliva) [reply model_building/model_building.py]: Done

File: model_building/model_building.py (hunk 9 above)
Comment (nahuel-coliva) [reply model_building/model_building.py]: Removed

File: model_building/wrapper_experiment_configuration.py (hunk 12 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: Removed

File: model_building/nnls_experiment_configuration.py (hunk 11 above)
Comment (nahuel-coliva) [reply model_building/nnls_experiment_configuration.py]: Removed and refactored

File: model_building/wrapper_experiment_configuration.py (hunk 13 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: Removed

File: model_building/wrapper_experiment_configuration.py (hunk 14 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: ^.^

File: model_building/wrapper_experiment_configuration.py (hunk 15 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: Removed as the other comments, in my defense I can assure you there were many other

File: model_building/model_building.py (hunk 10 above)
Comment (nahuel-coliva) [reply model_building/model_building.py]: Unindented

File: model_building/model_building.py (hunk 18 above)
Comment (nahuel-coliva) [reply model_building/model_building.py]: Refactored, here and some lines after

File: model_building/model_building.py (hunk 19 above)
Comment (nahuel-coliva) [reply model_building/model_building.py]: Added and modified

File: check.py (hunk 1 above)
Comment (brunoguindani) [reply check.py]: You didn't though 🤥

File: model_building/lr_ridge_experiment_configuration.py [hunk 20]
```diff
@@ -97,11 +97,14 @@ def print_model(self):
         initial_string = "LRRidge coefficients:\n"
         ret_string = initial_string
         coefficients = self._regressor.coef_
-        assert len(self._regressor.aml_features) == len(coefficients)
+
+        assert len(self.get_x_columns()) == len(coefficients) #len(self._regressor.aml_features) == len(coefficients)
```
Comment (brunoguindani) [inline model_building/lr_ridge_experiment_configuration.py]: Comment :)

File: model_building/lr_ridge_experiment_configuration.py [hunk 21]
```diff
@@ -97,11 +97,14 @@ def print_model(self):
...
+
         # Show coefficients in order of decresing absolute value
         idxs = np.argsort(np.abs(coefficients))[::-1]
+
         for i in idxs:
-            column = self._regressor.aml_features[i]
+            column = self.get_x_columns()[i]
```
Comment (brunoguindani) [inline model_building/lr_ridge_experiment_configuration.py]: Didn't you refactor this?

File: model_building/wrapper_experiment_configuration.py [hunk 22]
```diff
@@ -148,7 +155,8 @@ def get_x_columns(self):
         list of str:
             the columns used in the regression
         """
-        return self._wrapped_experiment_configuration.get_x_columns()
+        x_cols = super().get_x_columns() #self._wrapped_experiment_configuration.get_x_columns()
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: Comment!!!!11

File: model_building/wrapper_experiment_configuration.py [hunk 23]
```diff
@@ -599,6 +615,7 @@ def _train(self):
         """
         if self._hyperopt_trained:  # do not run Hyperopt again for the same exp.conf.
             SFSExperimentConfiguration._train(self)
+            self._logger.info("Hyperopt already run")
```
Comment (brunoguindani) [inline model_building/wrapper_experiment_configuration.py]: I'd put this in debug mode, not info, maybe with a slightly more detailed message

File: check.py (hunk 1 above)
Comment (nahuel-coliva) [reply check.py]: Now for real

File: model_building/lr_ridge_experiment_configuration.py (hunk 21 above)
Comment (nahuel-coliva) [reply model_building/lr_ridge_experiment_configuration.py]: That was in nnls, now here also

File: model_building/lr_ridge_experiment_configuration.py (hunk 20 above)
Comment (nahuel-coliva) [reply model_building/lr_ridge_experiment_configuration.py]: Removed

File: model_building/wrapper_experiment_configuration.py (hunk 22 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: Done

File: model_building/wrapper_experiment_configuration.py (hunk 23 above)
Comment (nahuel-coliva) [reply model_building/wrapper_experiment_configuration.py]: Now in debug, message slightly changed

**Review by brunoguindani — APPROVED** [review]
Great, thanks!

**Pull request merged**